src/engine/QueryPlanner.cpp

```diff
@@ -871,48 +852,74 @@ vector<QueryPlanner::SubtreePlan> QueryPlanner::seedWithScansAndText(
           plan._idsOfIncludedNodes |= (uint64_t(1) << i);
           auto& tree = *plan._qet.get();
           if (node._triple._p._iri == HAS_PREDICATE_PREDICATE) {
+            // TODO(schnelle): Handle ?p ql:has-prediacte ?p
             // Add a has relation scan instead of a normal IndexScan
             HasPredicateScan::ScanType scanType;
             if (isVariable(node._triple._s)) {
               scanType = HasPredicateScan::ScanType::FREE_S;
             } else if (isVariable(node._triple._o)) {
               scanType = HasPredicateScan::ScanType::FREE_O;
             }
-            std::shared_ptr<Operation> scan =
-                std::make_shared<HasPredicateScan>(_qec, scanType);
-            static_cast<HasPredicateScan*>(scan.get())
-                ->setSubject(node._triple._s);
-            static_cast<HasPredicateScan*>(scan.get())
-                ->setObject(node._triple._o);
+            auto scan = std::make_shared<HasPredicateScan>(_qec, scanType);
+            scan->setSubject(node._triple._s);
+            scan->setObject(node._triple._o);
             tree.setOperation(
                 QueryExecutionTree::OperationType::HAS_RELATION_SCAN, scan);
-            tree.setVariableColumns(static_cast<HasPredicateScan*>(scan.get())
-                                        ->getVariableColumns());
+            tree.setVariableColumns(scan->getVariableColumns());
+          } else if (isVariable(node._triple._s) &&
+                     isVariable(node._triple._o) &&
+                     node._triple._s == node._triple._o) {
+            if (isVariable(node._triple._p._iri)) {
+              LOG(WARN) << "Triple with one variable repeated 3 times"
+                        << std::endl;
+              return seeds;
```

Comment:
Shouldn't this be either a continue or an exception?

Reply:
Yeah I think you are right. I guess this might be legal so I'll make that clear in the exception. I just tried it though and the current return results in an a later exception of not finding a suitable execution tree but it's better to be explicit.